kubernetes/structures_stateful_set.go

```diff
@@ -0,0 +1,193 @@
+	spec, err := expandPersistentVolumeClaimSpec(in["spec"].([]interface{}))
+	if err != nil {
+		return corev1.PersistentVolumeClaim{}, err
+	}
+	pvc := corev1.PersistentVolumeClaim{
+		ObjectMeta: metadata,
+		Spec:       spec,
+	}
+	return pvc, nil
+}
+
+func expandStatefulSetSelectors(s []interface{}) (metav1.LabelSelector, error) {
+	obj := metav1.LabelSelector{}
+	if len(s) == 0 || s[0] == nil {
+		return obj, nil
+	}
+	in := s[0].(map[string]interface{})
+	log.Printf("[DEBUG] StatefulSet Selector: %#v", in)
+	if v, ok := in["match_labels"].(map[string]interface{}); ok {
+		log.Printf("[DEBUG] StatefulSet Selector MatchLabels: %#v", v)
+		ml := make(map[string]string)
+		for k, l := range v {
+			ml[k] = l.(string)
+			log.Printf("[DEBUG] StatefulSet Selector MatchLabel: %#v -> %#v", k, v)
+		}
+		obj.MatchLabels = ml
+	}
+	if v, ok := in["match_expressions"].([]interface{}); ok {
+		log.Printf("[DEBUG] StatefulSet Selector MatchExpressions: %#v", v)
+		me, err := expandMatchExpressions(v)
+		if err != nil {
+			return obj, err
+		}
+		obj.MatchExpressions = me
+	}
+	return obj, nil
+}
+
+func expandMatchExpressions(in []interface{}) ([]metav1.LabelSelectorRequirement, error) {
+	if len(in) == 0 {
+		return []metav1.LabelSelectorRequirement{}, nil
+	}
+	obj := make([]metav1.LabelSelectorRequirement, len(in))
+	for i, c := range in {
+		p := c.(map[string]interface{})
+		if v, ok := p["key"].(string); ok {
+			obj[i].Key = v
+		}
+		if v, ok := p["operator"].(metav1.LabelSelectorOperator); ok {
+			obj[i].Operator = v
+		}
+		if v, ok := p["values"].(*schema.Set); ok {
+			obj[i].Values = schemaSetToStringArray(v)
+		}
+	}
+	return obj, nil
+}
+
+// Flattners
+
+func flattenStatefulSetSpec(spec v1.StatefulSetSpec) ([]interface{}, error) {
+	att := make(map[string]interface{})
+
+	if spec.PodManagementPolicy != "" {
+		att["pod_management_policy"] = spec.PodManagementPolicy
+	}
+	if spec.Replicas != nil {
+		att["replicas"] = *spec.Replicas
+	}
+	if spec.RevisionHistoryLimit != nil {
+		att["revision_history_limit"] = *spec.RevisionHistoryLimit
+	}
+	if spec.Selector != nil {
+		att["selector"] = flattenLabelSelector(spec.Selector)
+	}
+	if spec.ServiceName != "" {
+		att["service_name"] = spec.ServiceName
+	}
+	template, err := flattenPodTemplateSpec(spec.Template)
+	if err != nil {
+		return []interface{}{att}, err
+	}
+	att["template"] = template
+	att["volume_claim_template"] = flattenPersistentVolumeClaim(spec.VolumeClaimTemplates)
+
+	return []interface{}{att}, nil
+}
+
+func flattenPodTemplateSpec(t corev1.PodTemplateSpec) ([]interface{}, error) {
+	template := make(map[string]interface{})
+
+	template["metadata"] = flattenMetadata(t.ObjectMeta)
+	spec, err := flattenPodSpec(t.Spec)
+	if err != nil {
+		return []interface{}{template}, err
+	}
+	template["spec"] = spec
+
+	return []interface{}{template}, nil
+}
+
+func flattenPersistentVolumeClaim(in []corev1.PersistentVolumeClaim) []interface{} {
+	pvcs := make([]interface{}, len(in))
```

Comment:
if you are going to use this form of `make` you shouldn't use `append`, or you can just add a `0` here for length: `make([]interface{}, 0, len(in))`

Reply:
Very good point! Seems a very common gotcha. Funny that I've noticed this exact issue in other PRs before and corrected it.